content/blog/handling-exceptions-in-specflow/index.md

```diff
@@ -0,0 +1,323 @@
+
+```csharp
+[Binding]
+class PersonSteps
+{
+    private readonly PersonRepository _people = new PersonRepository();
+    private string _actualName;
+
+    private readonly ErrorDriver _errorDriver;
+
+    public PersonSteps(ErrorDriver errorDriver)
+    {
+        _errorDriver = errorDriver;
+    }
+
+    /* Given steps omitted */
+
+    [When(@"I retrieve the person '(.*)'")]
+    public void WhenIRetrieveThePerson(string name)
+    {
+        _errorDriver.TryExecute(() =>
+            _actualName = _people.GetPersonByName(name)
+        );
+    }
+
+    /* Then step omitted */
+}
+```
+
+As you can see the new `ErrorDriver` class is injected into the `PersonSteps` class via [context injection](https://docs.specflow.org/projects/specflow/en/latest/Bindings/Context-Injection.html). The person specific `Given` and `Then` steps are unaltered. The `When` step no longer has a `try catch` block. Instead the action of retrieving a person is passed into the `TryExecute` method of the `ErrorDriver` class as a lambda. The `TryExecute` method will catch any exception as you'll see in moment. The `_actualException` private field is no longer used and has been removed.
+
+#### New generic ErrorSteps
+
+The `Then the error '<message>' should be raised` step and the `AfterScenario` hook are generic methods that can be reused for other features. I've moved these to a separate `ErrorSteps` class as shown below.
+
+```csharp
+[Binding]
+class ErrorSteps
+{
+    private readonly ErrorDriver _errorDriver;
+
+    public ErrorSteps(ErrorDriver errorDriver)
+    {
+        _errorDriver = errorDriver;
+    }
+
+    [Then(@"the error '(.*)' should be raised")]
+    public void ThenTheErrorShouldBeRaised(string expectedErrorMessage)
+    {
+        _errorDriver.AssertExceptionWasRaisedWithMessage(expectedErrorMessage);
+    }
+
+    [AfterScenario]
+    public void CheckForUnexpectedExceptionsAfterEachScenario()
+    {
+        _errorDriver.AssertNoUnexpectedExceptionsRaised();
+    }
+}
+```
+
+This class also receives the `ErrorDriver` class via context injection. It uses the available `Assert` methods to verify if an expected or unexpected error has occurred.
+
+#### The ErrorDriver class
+
+And here's the implementation of the `ErrorDriver` class.
+
+```csharp
+class ErrorDriver
+{
+    private readonly Queue<Exception> _exceptions = new Queue<Exception>();
+
+    public void TryExecute(Action act)
+    {
+        try
+        {
+            act();
+        }
+        catch (Exception ex)
+        {
+            Trace.WriteLine($"The following exception was caught while executing {act.Method.Name}: {ex}");
+            _exceptions.Enqueue(ex);
+        }
+    }
+
+    public void AssertExceptionWasRaisedWithMessage(string expectedErrorMessage)
+    {
+        Assert.IsTrue(_exceptions.Any(), $"No exception was raised but expected exception with message: {expectedErrorMessage}");
```

Comment:
Dit is een beetje op de inhoud van je oplossing? Wat vind jij van Asserts die niet in de steps zitten? Persoonlijk vind ik het iets minder overzichtelijk, maar ik ben benieuwd wat jij ervan vindt.

Reply:
Omdat mijn methode naam begint met Assert vind ik het hier geen probleem.

Als ik deze code in een package zou stoppen dan zou ik hier een (custom) exception gooien ipv de MSTest `Assert` class te gebruiken. Zodat het package niet afhankelijk is van MSTest / FluentAssertions / ... .